spec/controllers/reports_controller_spec.rb

```diff
@@ -0,0 +1,216 @@
+require "rails_helper"
+
+RSpec.describe ReportsController, type: :controller do
+  let!(:current_user){FactoryBot.create :user}
+  let!(:reports){FactoryBot.create_list :report, Settings.rspec.factory_reports_size, user_id: current_user.id}
+  let!(:comments){FactoryBot.create_list :comment, Settings.rspec.factory_comments_size, report_id: reports[0].id}
+  let!(:params) {FactoryBot.attributes_for :report}
+
+  context "When user does not log in" do
+    describe "GET #index" do
+      before {get :index}
+
+      it { expect(response).to redirect_to login_path }
+    end
+
+    describe "GET #new" do
+      before {get :new}
+
+      it { expect(response).to redirect_to login_path }
+    end
+
+    describe "GET #edit" do
+      before {get :edit, params: {id: reports[0].id}}
+
+      it { expect(response).to redirect_to login_path }
+    end
+
+    describe "GET #show" do
+      before {get :show, params: {id: reports[0].id}}
+
+      it { expect(response).to redirect_to login_path }
+    end
+  end
+
+  context "When user logged in" do
+    before {log_in current_user}
+
+    describe "GET #index" do
+      context "when do not have any params" do
+        before {get :index}
+
+        it { expect(assigns :reports).to eq(reports.reverse.slice(0, Settings.paginate.items_per_page)) }
```

Comment:
describe, context it là gì e nhỉ

Reply:
- describe dùng để nhóm các test case theo class, module, method hoặc theo action của controller. 

- context dùng để nhóm các điều kiện của example

- it được sử dụng để định nghĩa một "Ví dụ" hay nói cách khác là dùng để định nghĩa 1 test case. Chúng ta có thể mô tả cho it để test case được rõ ràng